Automattic/pocket-casts-android
# Search history: Data layer

Part of: #752

## Description
This PR
- Adds a new table `search_history` (+ migration) to the database
   - Adds unique indexes to restrict duplicate `search term`/ `podcast`/ `folder`/ `episode` search history item insertion
   - Includes `podcast`/ `folder`/ `episode` search history item details as embedded objects (with minimum details to render corresponding search history item) so that search history can be displayed for them even when there's no corresponding local `podcast`/ `folder`/ `episode` in the database.
- Adds `SearchHistoryDao` for search history item operations
   - findAll
   - insert (or replace)
   - delete
   - deleteAll


Note: Targets a feature branch to allow modifying the `search_history` table until the feature is ready.

## Testing Instructions

#### Tests
- Ensure that tests in `SearchHistoryDaoTest` cover important scenarios:
   - Single search history item insert and delete 
   - Inserting same `search term`/ `podcast`/ `folder`/ `episode` search history item replaces the previous item
   - Multiple unique `search term`/ `podcast`/ `folder`/ `episode` search history items are inserted
   - Last inserted/ replaced search history item is at the top of the search history
   - Delete all search history

#### Migration
- Install the app from the `main` branch
- Install the app from this PR branch
- Go to database inspector and notice that `search_history` table is created

#### Fresh Install 
- Clean install the app from this PR branch
- Go to database inspector and notice that `search_history` table is created

## Checklist
- If this is a user-facing change, I have added an entry in CHANGELOG.md
- All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...

- with different themes
- with a landscape orientation
- with the device set to have a large display and font size
- for accessibility with TalkBack
